Convert `t.Error(fmt.Sprintf(...))` to `t.Errorf(...)`.

d4l3k opened the pull request:
I noticed that golint warns about not using `errors.New(fmt.Sprintf(...))` and I noticed in my code I was using several instances of `t.Error(fmt.Sprintf(...))` in my tests so I implemented a lint check for it.

The pull request implements suggesting to replace `t.Error(fmt.Sprintf(...))` with `t.Errorf(...)` and the requisite tests.

dsymonds commented on lint.go:
```diff
@@ -1056,6 +1057,33 @@ func (f *file) lintErrorf() {
+
+	// testing.Error
+	testingTName := ""
+	f.walk(func(node ast.Node) bool {
```
This is too much duplicated code. Fold this into the previous walk

dsymonds commented on testdata/errorf_test.go:
```diff
@@ -0,0 +1,26 @@
+// Test for not using fmt.Errorf.
```
put this in the existing test file

dsymonds commented:
The general idea is fine, but we'll have to iterate on the implementation.

I also need you to sign the CLA: https://cla.developers.google.com/clas

d4l3k commented:
i signed and revised the code. Thoughts?

dsymonds commented on lint.go:
```diff
@@ -1038,22 +1038,38 @@ func (f *file) lintRanges() {
-		ce, ok = arg.(*ast.CallExpr)
-		if !ok || !isPkgDot(ce.Fun, "fmt", "Sprintf") {
-			return true
+		switch ce := node.(type) {
```
`ce` was an abbreviation of `CallExpr`, so it doesn't make sense to retain that name now. Use `v`.

dsymonds commented on lint.go:
```diff
@@ -1038,22 +1038,38 @@ func (f *file) lintRanges() {
 }
 
-// lintErrorf examines errors.New calls. It complains if its only argument is an fmt.Sprintf invocation.
+// lintErrorf examines errors.New and testing.Error calls. It complains if their only arguments are fmt.Sprintf invocations.
```
Switch back to singular in the second sentence.

dsymonds commented on lint.go:
```diff
@@ -1038,22 +1038,38 @@ func (f *file) lintRanges() {
-// lintErrorf examines errors.New calls. It complains if its only argument is an fmt.Sprintf invocation.
+// lintErrorf examines errors.New and testing.Error calls. It complains if their only arguments are fmt.Sprintf invocations.
 func (f *file) lintErrorf() {
+	testingTName := ""
```
```go
testingT := "" // name of the *testing.T arg last seen
```

dsymonds replied:
also add

```go
var testEnd token.Pos // end of test func
```

dsymonds replied:
actually, never mind. You shouldn't need either of these. see below.

dsymonds commented on lint.go:
```diff
@@ -1038,22 +1038,38 @@ func (f *file) lintRanges() {
+		switch ce := node.(type) {
+		case *ast.FuncDecl:
+			name := ce.Name.Name
+			if strings.HasPrefix(name, "Test") {
```
Let's lean on the typechecker a bit more here. Instead of tracking whether we're in a test func, use the typechecker when you see `t.Error` to identify the type of `t`. Use `f.pkg.typeOf` and `f.pkg.isNamedType` (after checking for types.Pointer). Then this walk only needs to inspect `CallExpr`s again.

dsymonds commented on lint.go:
```diff
@@ -1038,22 +1038,38 @@ func (f *file) lintRanges() {
+			if !ok || !isPkgDot(ce.Fun, "fmt", "Sprintf") {
+				return true
+			}
+			err := "should replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)"
```
don't duplicate this whole string just to change one piece. Use f.render on the right bit of the AST (probably a SelectorExpr).

dsymonds commented on testdata/errorf.go:
```diff
@@ -21,4 +22,19 @@ func f(x int) error {
+func TestF(t *testing.T) error {
+	x := 1
+	if x > 10 {
+		return t.Error(fmt.Sprintf("something %d", x)) // MATCH /should replace.*t\.Error\(fmt\.Sprintf\(\.\.\.\)\).*t\.Errorf\(\.\.\.\)/
```
no need to be this specific for this one.

dsymonds replied:
by specific, I mean that the regexp is tighter than it needs to be. but now that I think about it more, this is fine as is.

d4l3k replied:
Fair enough.

dsymonds commented:
Don't forget to remove `testdata/errorf_test.go` entirely.

d4l3k commented:
Oops. I removed it but forgot to check the change in. I implemented most of the changes but I wasn't sure what you meant by "no need to be this specific for this one.".

dsymonds commented on lint.go:
```diff
@@ -1038,14 +1038,17 @@ func (f *file) lintRanges() {
 func (f *file) lintErrorf() {
 	f.walk(func(node ast.Node) bool {
 		ce, ok := node.(*ast.CallExpr)
 		if !ok {
```
move the `len(ce.Args) != 1` check up to here.

dsymonds commented on lint.go:
```diff
@@ -1038,14 +1038,17 @@ func (f *file) lintRanges() {
 			return true
 		}
-		if !isPkgDot(ce.Fun, "errors", "New") || len(ce.Args) != 1 {
+		isErrorsNew := isPkgDot(ce.Fun, "errors", "New")
```
This still isn't quite right. It's promoting the particular case too far.

How about something more like this:
```go
check := false
if isPkgDot(ce.Fun, "errors", "New") {
  check = true
} else if se, ok := ce.Fun.(*ast.SelectorExpr); ok {
  check = se.Sel.Name == "Error" && f.pkg.typeOf(se.X).String() == "*testing.T"
}
if !check {
  return true
}
```

dsymonds replied:
sorry, don't make this change just yet...

dsymonds commented on lint.go:
```diff
@@ -1053,7 +1056,11 @@ func (f *file) lintErrorf() {
 			return true
 		}
-		f.errorf(node, 1, category("errors"), "should replace errors.New(fmt.Sprintf(...)) with fmt.Errorf(...)")
+		errorfPrefix := "fmt"
```
ugh, I see. I didn't think this through enough. So we need the bools from above to make this switch correctly. okay, leave the previous code as it was (except for hoisting the `len(ce.Args) != 1` check, since it is so cheap).

d4l3k commented:
Any other thoughts?

dsymonds commented on lint.go:
```diff
@@ -1038,14 +1038,17 @@ func (f *file) lintRanges() {
+		isErrorsNew := isPkgDot(ce.Fun, "errors", "New")
+		se, ok := ce.Fun.(*ast.SelectorExpr)
+		isTestingError := ok && se.Sel.Name == "Error" && f.pkg.typeOf(se.X).String() == "*testing.T"
+		if !(isErrorsNew || isTestingError) {
```
if !isErrorsNew && !isTestingError {

dsymonds commented:
Okay, just one last small change and then I'll merge this.

d4l3k commented:
Changed. Thanks for walking me through this!
